entry/csp.js

```diff
@@ -0,0 +1,102 @@
+import { createHash } from 'crypto';
+import { parse } from 'node-html-parser';
+import { URL } from 'url';
+
+const scriptTypeIgnoreList = ['application/json', 'application/ld+json'];
+
+const defaultBaseName = 'http://relative-url';
+
+const hashScriptContents = (scriptContents) =>
+  createHash('sha256').update(scriptContents).digest('base64');
+
+export default function createCSPHandler({ extraHosts = [] } = {}) {
+  let tagReturned = false;
+  const hosts = new Set();
+  const shas = new Set();
+
+  const addScriptContents = (contents) => {
+    shas.add(hashScriptContents(contents));
+  };
+
+  const addScriptUrl = (src) => {
+    const { origin } = new URL(src, defaultBaseName);
+
+    if (origin !== defaultBaseName) {
+      hosts.add(origin);
+    }
+  };
+
+  extraHosts.forEach((host) => addScriptUrl(host));
+
+  const processScriptNode = (scriptNode) => {
+    const src = scriptNode.getAttribute('src');
+
+    if (src) {
+      addScriptUrl(src);
+    } else if (
+      !scriptTypeIgnoreList.includes(scriptNode.getAttribute('type'))
+    ) {
+      addScriptContents(scriptNode.firstChild.rawText);
+    }
+  };
+
+  const registerScript = (script) => {
+    if (tagReturned) {
+      throw new Error(
+        `Unable to register script. Content Security Policy already sent. Try registering scripts before calling flushHeadTags. Script: ${script.substr(
+          0,
+          30,
+        )}`,
+      );
+    }
+
+    const root = parse(script, { script: true });
+
+    if (!root.valid) {
+      console.error(`Invalid script passed to 'csp.registerScript'`);
+    }
+
+    root.querySelectorAll('script').forEach(processScriptNode);
+  };
+
+  const createCSPTag = () => {
+    tagReturned = true;
+
+    const inlineCspShas = [];
+
+    for (const sha of shas.values()) {
+      inlineCspShas.push(`'sha256-${sha}'`);
+    }
+
+    const scriptSrcPolicy = [
+      'script-src',
+      `'self'`,
+      ...hosts.values(),
+      ...inlineCspShas,
+    ].join(' ');
+
+    return `<meta http-equiv="Content-Security-Policy" content="${scriptSrcPolicy};">`;
+  };
+
+  const handleHtml = (html) => {
+    const root = parse(html, { script: true });
+
+    if (!root.valid) {
+      throw new Error(
+        `Unable to parse HTML in order to create CSP tag. Check the following output of renderDocument for invalid HTML.\n${html}`,
```

Comment:
I hear you, but how else can they debug the issue?

Reply:
Could we do the same as the script error message, put a reasonable limit on the HTML. Just enough to know what the HTML roughtly looks like but by pointing them to renderDocument they know if they need more info they can log out their renderDocument function.
```
html.length > 400 ? `${html.subStr(0, 200)}...` : html;
```
The limit can be quite high because it's just about not completely spamming the console out.